Options.py

```diff
@@ -466,6 +465,94 @@ def __eq__(self, other: typing.Any):
 
 
+class Bosses(TextChoice):
+    """Generic boss shuffle option that supports plando. Format expected is 'boss-location;boss-location;shuffle_type'
+    if shuffle_type is not provided in the string, will default to null. Must override can_place_boss, which
+    passes a plando boss and location. Check if the placement is valid for your game here."""
+    bosses: set = set()
+    locations: set = set()
+
+    @classmethod
+    def from_text(cls, text: str):
+        # set all of our text to lower case for name checking
+        text = text.lower()
+        if text == "random":
+            import random
+            return cls(random.choice(list(cls.options.values())))
+        for option_name, value in cls.options.items():
+            if option_name == text:
+                return cls(value)
+        cls.bosses = {boss_name.lower() for boss_name in cls.bosses}
+        cls.locations = {boss_location.lower() for boss_location in cls.locations}
+        options = text.split(";")
+
+        # since plando exists in the option verify the plando values given are valid
+        cls.validate_plando_bosses(options)
+        return cls.get_shuffle_type(options)
+
+    @classmethod
+    def get_shuffle_type(cls, options: typing.List[str]):
+        # find out what type of boss shuffle we should use for placing bosses after plando
+        # and add as a string to look nice in the spoiler
+        if "random" in options:
+            import random
+            shuffle = random.choice(list(cls.options))
+            options.remove("random")
+            options = ";".join(options) + ";" + shuffle
+            boss_class = cls(options)
+        else:
+            for option in options:
+                if option in cls.options:
+                    options = ";".join(options)
+                    break
+            else:
+                options = ";".join(options) + ";none"
+            boss_class = cls(options)
+        return boss_class
+
+    @classmethod
+    def validate_plando_bosses(cls, options: typing.List[str]) -> None:
+        for option in options:
+            # check if a shuffle type was provided in the incorrect location
+            if option == "random" or option in cls.options:
+                if option != options[-1]:
+                    raise ValueError(F"{option} option must be at the end of the boss_shuffle options!")
+            elif "-" in option:
+                location, boss = option.split("-")
+                if not cls.valid_boss_name(boss):
+                    raise ValueError(F"{boss.title()} is not a valid boss name for location {location.title()}.")
+                if not cls.valid_location_name(location):
+                    raise ValueError(f"{location.title()} is not a valid boss location name for boss {boss.title()}.")
+                if not cls.can_place_boss(boss, location):
+                    raise ValueError(f"{location.title()} is not a valid location for {boss} to be placed.")
```

Comment:
The first two may be misleading. If boss is not valid at all (in no location), it should not say "for location ..." because that sounds like the location is wrong, not the boss.

The last one does not do .title()